pydantic/_internal/_generate_schema.py

```diff
@@ -267,12 +267,36 @@ def _add_custom_serialization_from_json_encoders(
     return schema
 
 
+TypesNamespace = dict[str, Any] | None
+
+
+class TypesNamespaceStack:
```

Comment:
This is very closely related to/inspired by the `pydantic._internal._config.ConfigWrapperStack`